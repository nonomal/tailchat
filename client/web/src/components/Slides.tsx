import { Carousel } from 'antd';
import type { CarouselProps, CarouselRef } from 'antd/lib/carousel';
import React, { useImperativeHandle, useRef } from 'react';

import './Slides.less';

export interface SlidesRef {
  next: () => void;
  prev: () => void;
}

/**
 * 类似于走马灯, 但是用于分步操作
 *
 * 会随内容的变化动态改变高度
 *
 * Reference: https://react-slick.neostack.com/docs/api
 */
export const Slides = React.forwardRef<SlidesRef, CarouselProps>(
  (props, ref) => {
    const carouselRef = useRef<CarouselRef>(null);

    useImperativeHandle(
      ref,
      () => ({
        next() {
          carouselRef.current?.next();
        },
        prev() {
          carouselRef.current?.prev();
        },
      }),
      []
    );

    return (
      <Carousel
        className="slides"
        ref={carouselRef}
        {...props}
        dots={false}
        swipe={false}
        adaptiveHeight={true}
        infinite={false}
        beforeChange={(current, next) => {
          console.log(current, next, carouselRef.current?.innerSlider);
        }}
      >
        {props.children}
      </Carousel>
    );
  }
);
Slides.displayName = 'Slides';
