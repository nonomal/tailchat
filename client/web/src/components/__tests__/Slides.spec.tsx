import { render } from '@testing-library/react';
import React from 'react';
import { Slides } from '../Slides';

describe('Slides', () => {
  test('render', () => {
    const wrapper = render(
      <Slides>
        <div>
          <h1>1</h1>
        </div>
        <div>
          <h1>2</h1>
        </div>
        <div>
          <h1>3</h1>
        </div>
      </Slides>
    );
    expect(wrapper.container).toMatchSnapshot();
  });
});
