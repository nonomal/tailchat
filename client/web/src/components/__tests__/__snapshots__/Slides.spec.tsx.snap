// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Slides render 1`] = `
<div>
  <div
    class="ant-carousel"
  >
    <div
      class="slick-slider slides slick-initialized"
      dir="ltr"
    >
      <div
        class="slick-list"
        style="height: 0px;"
      >
        <div
          class="slick-track"
          style="opacity: 1; transform: translate3d(0px, 0px, 0px);"
        >
          <div
            aria-hidden="false"
            class="slick-slide slick-active slick-current"
            data-index="0"
            style="outline: none; width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                style="width: 100%; display: inline-block;"
                tabindex="-1"
              >
                <h1>
                  1
                </h1>
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            class="slick-slide"
            data-index="1"
            style="outline: none; width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                style="width: 100%; display: inline-block;"
                tabindex="-1"
              >
                <h1>
                  2
                </h1>
              </div>
            </div>
          </div>
          <div
            aria-hidden="true"
            class="slick-slide"
            data-index="2"
            style="outline: none; width: 0px;"
            tabindex="-1"
          >
            <div>
              <div
                style="width: 100%; display: inline-block;"
                tabindex="-1"
              >
                <h1>
                  3
                </h1>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
